lib/ruby_indexer/lib/ruby_indexer/prefix_tree.rb

```diff
@@ -138,10 +138,11 @@ def initialize(key, value, parent = nil)
       #: -> Array[Value]
       def collect
         result = []
-        result << @value if @leaf
+        stack = [self]
 
-        @children.each_value do |node|
-          result.concat(node.collect)
+        while (node = stack.pop)
+          result.prepend(node.value) if node.leaf
+          stack.concat(node.children.values)
```

Comment:
Another optimization here would be to not push anything to the stack if the node is leaf:
```suggestion
          if node.leaf
            result.prepend(node.value)
          else
            stack.concat(node.children.values)
          end
```

Reply:
I don't think we can do this because leaf nodes can also have children. For example, if we have `ActiveRecord` and `ActiveRecord::Base` in the prefix tree, the node for `ActiveRecord` will both have a value associated with it (the module entry) and have children that take us to `ActiveRecord::Base`.

Do you agree or am I misunderstanding your suggestion?